File: modules/fsrs-native/android/src/main/java/expo/modules/fsrsnative/FsrsNativeModule.kt
package expo.modules.fsrsnative

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class FsrsNativeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("FsrsNative")

    // 工具鏈探針（與 iOS 對齊回 42；確認 Expo 模組載入）。
    Function("ping") {
      42
    }

    // 用複習歷史訓練 FSRS 參數。計算密集 → AsyncFunction（背景執行緒）。
    // 輸入/輸出皆 JSON 字串，透過 JNI 呼叫 Rust。
    AsyncFunction("optimize") { inputJson: String ->
      FsrsNativeRust.optimize(inputJson)
    }
  }
}

File: modules/fsrs-native/ios/FsrsNativeModule.swift
import ExpoModulesCore
import FsrsMobile // Clang module（FsrsMobile.xcframework）：fsrs_native_ping / fsrs_optimize / fsrs_free

public class FsrsNativeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("FsrsNative")

    // 工具鏈探針（同步、極輕量）。
    Function("ping") { () -> Int in
      Int(fsrs_native_ping())
    }

    // 用複習歷史訓練 FSRS 參數。計算密集 → AsyncFunction（背景執行緒）。
    // 輸入/輸出皆為 JSON 字串，避免跨橋傳遞複雜結構。
    AsyncFunction("optimize") { (inputJson: String) -> String in
      guard let result = fsrs_optimize(inputJson) else {
        return "{\"ok\":false,\"w\":[],\"error\":\"null result\"}"
      }
      defer { fsrs_free(result) }
      return String(cString: result)
    }
  }
}
